### PR TITLE
Forms deletion

### LINKED ISSUE #14272: Support deletion of the Forms through DeleteResource gRPC
**Description**

With the epic #12874, deploying Forms to Zeebe will be possible. We now support deleting resources through DeleteResource gRPC and it needs to be updated for Form resource deletion. Details of the implementation will be provided later.

### PR DESCRIPTION
**Pull request by nicpuppa**
## Description

This PR enable the deletion of a form. 

## Related issues

closes #14272 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: engine/src/test/java/io/camunda/zeebe/engine/state/deployment/FormStateTest.java [hunk 1]
```diff
@@ -96,6 +96,68 @@ void shouldReturnEmptyIfNoFormIsDeployedForFormKey() {
...
+
+    // then
+    assertThat(formState.findLatestFormById(form.getFormIdBuffer(), form.getTenantId())).isEmpty();
+  }
+
+  @Test
+  void shouldNotFindVersion2AsLatestFormAfterDeletion() {
+    // given
+    final var formV1 = sampleFormRecord();
+    final var formV2 = sampleFormRecord(2, formV1.getFormKey());
```
Comment (remcowesterhoud) [inline engine/src/test/java/io/camunda/zeebe/engine/state/deployment/FormStateTest.java:119]: ❌ You are giving `formV2` the same key as `formV1`. This is not a realistic case as a newly deployed form will always have a new key as well. Only the `formId` should be the same.

This doesn't impact the test as it is currently. But if you would try to find `formV1` by key you would not get any result.

Comment (remcowesterhoud) [reply engine/src/test/java/io/camunda/zeebe/engine/state/deployment/FormStateTest.java:119]: Same for the tests below this one.

**Review by remcowesterhoud — CHANGES_REQUESTED** [review]
Overall great work @nicpuppa 🏆 I have a few comments, so please have a look at those.


I'd also like to raise 2 additional things:
1. We must make sure we also update [the documentation](https://docs.camunda.io/docs/next/components/concepts/resource-deletion/)
2. Does Tasklist have an issue for this on their radar? They must make sure to consume the newly exported event and delete the Forms on their side as well. Especially considering they will show the latest versions of Forms in case of a Form Start Event. It would be weird if a user deleted the latest version from Zeebe, but Tasklist still shows it.
It's also interesting to know what would happens with existing User Tasks if the form related to them is deleted. But that's for Tasklist to decide.

File: engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbFormState.java [hunk 2]
```diff
@@ -83,6 +83,22 @@ public void storeFormRecord(final FormRecord record) {
...
+    formVersion.wrapLong(record.getVersion());
+
+    formsByKey.deleteExisting(tenantAwareFormKey);
+    formByIdAndVersionColumnFamily.deleteExisting(tenantAwareIdAndVersionKey);
+
+    versionManager.deleteResourceVersion(
+        record.getFormId(), record.getVersion(), record.getTenantId());
+    formByTenantAndIdCache.remove(
+        new TenantIdAndFormId(record.getTenantId(), record.getFormIdBuffer()));
+  }
```
[the remark this answers is not in the document]
Comment (nicpuppa) [reply engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbFormState.java:100]: I wasn't aware of this decision. I will try to implement the atomic operation for the form state. And find a way to test it. If I got stuck somewhere I will let you know. Thank you 💪

File: engine/src/test/java/io/camunda/zeebe/engine/state/deployment/FormStateTest.java (hunk 1 above)
Comment (nicpuppa) [reply engine/src/test/java/io/camunda/zeebe/engine/state/deployment/FormStateTest.java:119]: You are right, but I don't care about the `formKey` in this tests. But I will update the form key in order to represent a realistic case 👍

**Comment by nicpuppa** [comment]
Hey @remcowesterhoud. Thanks for the review 🙏 Lot of good suggestions as usual.

> 1. We must make sure we also update [the documentation](https://docs.camunda.io/docs/next/components/concepts/resource-deletion/)

I will create an issue on the doc for this. As form deletion is not yet officially supported

> 2\. Does Tasklist have an issue for this on their radar? They must make sure to consume the newly exported event and delete the Forms on their side as well. Especially considering they will show the latest versions of Forms in case of a Form Start Event. It would be weird if a user deleted the latest version from Zeebe, but Tasklist still shows it.
It's also interesting to know what would happens with existing User Tasks if the form related to them is deleted. But that's for Tasklist to decide.

I'm not aware of that, maybe @berkaycanbc has more vision about the future plan. As far as I know, at the moment, only zeebe support the deletion of a form

**Comment by berkaycanbc** [comment]
> 2\. Does Tasklist have an issue for this on their radar? They must make sure to consume the newly exported event and delete the Forms on their side as well. Especially considering they will show the latest versions of Forms in case of a Form Start Event. It would be weird if a user deleted the latest version from Zeebe, but Tasklist still shows it.
It's also interesting to know what would happens with existing User Tasks if the form related to them is deleted. But that's for Tasklist to decide.

No, we should communicate this. I will check in the project channel if they will be able to implement it for 8.4. Otherwise, we should hold this one.

> It's also interesting to know what would happens with existing User Tasks if the form related to them is deleted.

I expect we still show the previous latest one since we export the `fromKey` while creating the job for user task. But I agree, this is for Tasklist to decide anyways.

@remcowesterhoud @nicpuppa

**Review by remcowesterhoud — CHANGES_REQUESTED** [review]
@nicpuppa I just got triggered by a slack thread. We should introduce another test case where we verify version numbers are not reused.

E.g.
1. Deploy form `foo` (becomes v1)
2. Delete form `foo` v1
3. Deploy form `foo`
4. Verify it gets v2

**Review by berkaycanbc — COMMENTED** [review]
I also took a look. No additional comments. 👍

**Comment by nicpuppa** [comment]
Hey @remcowesterhoud I applied all your suggestions. Let me know your thoughts, especially on the atomic operation topic.

Also please @berkaycanbc have another look if you want!

File: engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbFormState.java [hunk 3]
```diff
@@ -70,35 +70,62 @@ public DbFormState(
...
+        record.getFormId(), record.getVersion(), record.getTenantId());
   }
 
   @Override
-  public void deleteForm(final FormRecord record) {
+  public void updateLatestFormCache(final FormRecord record) {
+    formByTenantAndIdCache.put(
+        new TenantIdAndFormId(record.getTenantId(), record.getFormIdBuffer()),
+        dbPersistedForm.copy());
+  }
```
[the remark this answers is not in the document]
Comment (remcowesterhoud) [reply engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbFormState.java:100]: ❌ The way it is now there is no wrapping of the record. This means if you call this method without the `storeFormInFormColumn` method it won't behave as expected and might cache the wrong thing.

This makes my argument to combine the two methods even stronger 😄

**Review by remcowesterhoud — CHANGES_REQUESTED** [review]
I like it @nicpuppa 👍 I do have a bunch of comments though 😄 

💭 ❓ What do you think about renaming the `FormStateMultiTenantTest` and `FormStateTest`? They don't really test the state directly anymore, but they now test the event applier. My proposal would be having `FormCreatedApplierTest` and `FormDeletedApplierTest` and separating the tests accordingly. The Multi-Tenancy test can just be part of those classes instead of a separate class.

File: engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbFormState.java [hunk 4]
```diff
@@ -70,35 +70,62 @@ public DbFormState(
   }
 
   @Override
-  public void storeFormRecord(final FormRecord record) {
+  public void storeFormInFormColumn(final FormRecord record) {
```
[the remark this answers is not in the document]
Comment (nicpuppa) [reply engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbFormState.java:73]: Do you like the name of those methods ? @remcowesterhoud

[the remark this answers is not in the document]
Comment (remcowesterhoud) [reply engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbFormState.java:73]: I'd make it `storeFormInFormColumnFamily` personally.

File: engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbFormState.java [hunk 5]
```diff
@@ -70,35 +70,62 @@ public DbFormState(
...
     tenantIdKey.wrapString(record.getTenantId());
     dbFormKey.wrapLong(record.getFormKey());
+    dbPersistedForm.wrap(record);
+    formsByKey.upsert(tenantAwareFormKey, dbPersistedForm);
```
[the remark this answers is not in the document]
Comment (nicpuppa) [reply engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbFormState.java:77]: No particular reason. Just saw that is used also in other resources states (processes and decisions). The only things that came to my mind is that we added later the check on duplicate forms.

**Comment by nicpuppa** [comment]
> 💭 ❓ What do you think about renaming the `FormStateMultiTenantTest` and `FormStateTest`? They don't really test the state directly anymore, but they now test the event applier. My proposal would be having `FormCreatedApplierTest` and `FormDeletedApplierTest` and separating the tests accordingly. The Multi-Tenancy test can just be part of those classes instead of a separate class.

It is not fully true, some tests still verify single state method. We can separate those tests by creating 2 new tests classes (`FormCreatedApplierTest` and `FormDeletedApplierTest`) and move there the tests that verify the whole chain of creation/deletion of a form.

@remcowesterhoud

File: engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbFormState.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (remcowesterhoud) [reply engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbFormState.java:77]: I think it actually has to do with distribution and a partition potentially receiving the distributed command twice. It's fine as-is 🙂

[the remark this answers is not in the document]
Comment (nicpuppa) [reply engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbFormState.java:77]: Make sense 😄

**Comment by nicpuppa** [comment]
Please @remcowesterhoud have another look 👀

**Review by remcowesterhoud — APPROVED** [review]
LGTM

**Comment by nicpuppa** [comment]
bors merge

**Pull request merged**